streaming/sender.c

```diff
@@ -0,0 +1,697 @@
+#ifdef ENABLE_HTTPS
+    if (s->host->ssl.conn && !s->host->stream_ssl.flags) {
+        int desired = sizeof(s->read_buffer) - s->read_len - 1;
+        ret = SSL_read(s->host->ssl.conn, s->read_buffer, desired);
+        if (ret > 0 ) {
+            s->read_len += ret;
+            return;
+        }
+        int sslerrno = SSL_get_error(s->host->ssl.conn, desired);
+        if (sslerrno == SSL_ERROR_WANT_READ || sslerrno == SSL_ERROR_WANT_WRITE)
+            return;
+        u_long err;
+        char buf[256];
+        while ((err = ERR_get_error()) != 0) {
+            ERR_error_string_n(err, buf, sizeof(buf));
+            error("STREAM %s [send to %s] ssl error: %s", s->host->hostname, s->connected_to, buf);
+        }
+        error("Restarting connection");
+        rrdpush_sender_thread_close_socket(s->host);
+        return;
+    }
+#endif
+    ret = recv(s->host->rrdpush_sender_socket, s->read_buffer + s->read_len, sizeof(s->read_buffer) - s->read_len - 1,
+               MSG_DONTWAIT);
+    if (ret>=0) {
+        s->read_len += ret;
+        return;
+    }
+    if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR)
+        return;
+    error("STREAM %s [send to %s]: error during read (%d). Restarting connection", s->host->hostname, s->connected_to,
+          ret);
+    rrdpush_sender_thread_close_socket(s->host);
+}
+
+// This is just a placeholder until the gap filling state machine is inserted
+void execute_commands(struct sender_state *s) {
+    char *start = s->read_buffer, *end = &s->read_buffer[s->read_len], *newline;
+    *end = 0;
+    while( start<end && (newline=strchr(start, '\n')) ) {
+        *newline = 0;
+        info("STREAM %s [send to %s] received command over connection: %s", s->host->hostname, s->connected_to, start);
+        start = newline+1;
+    }
+    if (start<end) {
+        memcpy( s->read_buffer, start, end-start);
+        s->read_len = end-start;
+    }
+}
+
+
+static void rrdpush_sender_thread_cleanup_callback(void *ptr) {
+    RRDHOST *host = (RRDHOST *)ptr;
+
+    rrdhost_wrlock(host);
+
+    info("STREAM %s [send]: sending thread cleans up...", host->hostname);
+
+    rrdpush_sender_thread_close_socket(host);
+
+    // close the pipe
+    if(host->rrdpush_sender_pipe[PIPE_READ] != -1) {
+        close(host->rrdpush_sender_pipe[PIPE_READ]);
+        host->rrdpush_sender_pipe[PIPE_READ] = -1;
+    }
+
+    if(host->rrdpush_sender_pipe[PIPE_WRITE] != -1) {
+        close(host->rrdpush_sender_pipe[PIPE_WRITE]);
+        host->rrdpush_sender_pipe[PIPE_WRITE] = -1;
+    }
+
+    if(!host->rrdpush_sender_join) {
+        info("STREAM %s [send]: sending thread detaches itself.", host->hostname);
+        netdata_thread_detach(netdata_thread_self());
+    }
+
+    host->rrdpush_sender_spawn = 0;
+
+    info("STREAM %s [send]: sending thread now exits.", host->hostname);
+
+    rrdhost_unlock(host);
+}
+
+void sender_init(struct sender_state *s, RRDHOST *parent) {
+    memset(s, 0, sizeof(*s));
```

Comment:
A suggestion, instead to use `mallocz`, you could use `callocz` and remove this `memset`.

Reply:
This is true, although as it stands this procedure guarantees the initialized state of the structure without any preconditions.